File: dart/test/authdog_client_test.dart
import 'dart:convert';
import 'package:test/test.dart';
import 'package:http/http.dart' as http;
import 'package:mocktail/mocktail.dart';

import 'package:authdog_sdk/authdog_client.dart';
import 'package:authdog_sdk/exceptions.dart';

class MockHttpClient extends Mock implements http.Client {}

void main() {
  setUpAll(() {
    registerFallbackValue(Uri.parse('https://example.com'));
    registerFallbackValue(http.Request('GET', Uri.parse('https://example.com')));
  });

  group('AuthdogClient', () {
    late MockHttpClient mockClient;
    late AuthdogClient client;

    setUp(() {
      mockClient = MockHttpClient();
      client = AuthdogClient(AuthdogClientConfig(
        baseUrl: 'https://api.authdog.com',
        httpClient: mockClient,
      ));
    });

    test('getUserInfo success', () async {
      final responseBody = json.encode({
        'meta': {'code': 200, 'message': 'OK'},
        'session': {'remainingSeconds': 3600},
        'user': {'id': '123', 'externalId': 'e', 'userName': 'u', 'displayName': 'd', 'nickName': null, 'profileUrl': null, 'title': null, 'userType': null, 'preferredLanguage': null, 'locale': 'en', 'timezone': null, 'active': true, 'names': {'id': 'n', 'formatted': 'd', 'familyName': 'ln', 'givenName': 'gn', 'middleName': null, 'honorificPrefix': null, 'honorificSuffix': null}, 'photos': [], 'phoneNumbers': [], 'addresses': [], 'emails': [], 'verifications': [], 'provider': 'p', 'createdAt': '2023-01-01T00:00:00Z', 'updatedAt': '2023-01-01T00:00:00Z', 'environmentId': 'env'}
      });

      when(() => mockClient.get(
            any(that: isA<Uri>()),
            headers: any(named: 'headers'),
          )).thenAnswer((_) async => http.Response(responseBody, 200, headers: {'content-type': 'application/json'}));

      final result = await client.getUserInfo('token');
      expect(result.user.id, equals('123'));
      expect(result.meta.code, equals(200));
    });

    test('getUserInfo 401 throws AuthenticationException', () async {
      when(() => mockClient.get(any(), headers: any(named: 'headers')))
          .thenAnswer((_) async => http.Response('', 401));

      expect(() => client.getUserInfo('bad'), throwsA(isA<AuthenticationException>()));
    });

    test('getUserInfo 500 GraphQL error', () async {
      final body = json.encode({'error': 'GraphQL query failed'});
      when(() => mockClient.get(any(), headers: any(named: 'headers')))
          .thenAnswer((_) async => http.Response(body, 500));

      expect(() => client.getUserInfo('t'), throwsA(predicate((e) => e is ApiException && e.toString().contains('GraphQL query failed'))));
    });

    test('getUserInfo 500 fetch error', () async {
      final body = json.encode({'error': 'Failed to fetch user info'});
      when(() => mockClient.get(any(), headers: any(named: 'headers')))
          .thenAnswer((_) async => http.Response(body, 500));

      expect(() => client.getUserInfo('t'), throwsA(predicate((e) => e is ApiException && e.toString().contains('Failed to fetch user info'))));
    });

    test('getUserInfo non-2xx throws ApiException', () async {
      when(() => mockClient.get(any(), headers: any(named: 'headers')))
          .thenAnswer((_) async => http.Response('Bad', 400));

      expect(() => client.getUserInfo('t'), throwsA(isA<ApiException>()));
    });

    test('close does not throw', () {
      expect(() => client.close(), returnsNormally);
    });
  });
}

File: planned/kotlin/src/test/kotlin/com/authdog/sdk/AuthdogClientTest.kt
package com.authdog.sdk

import io.mockk.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.net.URI
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class AuthdogClientTest {

    private lateinit var mockHttpClient: HttpClient
    private lateinit var mockResponse: HttpResponse<String>
    private lateinit var client: AuthdogClient

    @BeforeEach
    fun setUp() {
        mockHttpClient = mockk()
        mockResponse = mockk()
        client = AuthdogClient(
            AuthdogClientConfig(
                baseUrl = "https://api.authdog.com",
                apiKey = "test-api-key",
                httpClient = mockHttpClient
            )
        )
    }

    @AfterEach
    fun tearDown() {
        clearAllMocks()
    }

    @Test
    fun `constructor with baseUrl sets correct configuration`() = runTest {
        val testClient = AuthdogClient(
            AuthdogClientConfig(
                baseUrl = "https://api.authdog.com"
            )
        )
        
        assertNotNull(testClient)
        testClient.close()
    }

    @Test
    fun `constructor with trailing slash removes trailing slash`() = runTest {
        val testClient = AuthdogClient(
            AuthdogClientConfig(
                baseUrl = "https://api.authdog.com/"
            )
        )
        
        assertNotNull(testClient)
        testClient.close()
    }

    @Test
    fun `constructor with apiKey sets apiKey`() = runTest {
        val testClient = AuthdogClient(
            AuthdogClientConfig(
                baseUrl = "https://api.authdog.com",
                apiKey = "test-key"
            )
        )
        
        assertNotNull(testClient)
        testClient.close()
    }

    @Test
    fun `constructor with custom httpClient uses provided client`() = runTest {
        val customClient = mockk<HttpClient>()
        val testClient = AuthdogClient(
            AuthdogClientConfig(
                baseUrl = "https://api.authdog.com",
                httpClient = customClient
            )
        )
        
        assertNotNull(testClient)
        testClient.close()
    }

    @Test
    fun `getUserInfo with valid token returns user info`() = runTest {
        val mockResponseBody = """
        {
            "meta": {"code": 200, "message": "Success"},
            "session": {"remainingSeconds": 3600},
            "user": {
                "id": "user123",
                "externalId": "ext123",
                "userName": "testuser",
                "displayName": "Test User",
                "locale": "en-US",
                "active": true,
                "names": {
                    "id": "name123",
                    "familyName": "User",
                    "givenName": "Test"
                },
                "photos": [],
                "phoneNumbers": [],
                "addresses": [],
                "emails": [],
                "verifications": [],
                "provider": "test",
                "createdAt": "2023-01-01T00:00:00Z",
                "updatedAt": "2023-01-01T00:00:00Z",
                "environmentId": "env123"
            }
        }
        """.trimIndent()

        every { mockResponse.statusCode() } returns 200
        every { mockResponse.body() } returns mockResponseBody
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        val result = client.getUserInfo("valid-token")

        assertEquals("user123", result.user.id)
        assertEquals("testuser", result.user.userName)
        assertEquals("Test User", result.user.displayName)
        assertEquals(200, result.meta.code)
        assertEquals(3600, result.session.remainingSeconds)
    }

    @Test
    fun `getUserInfo with unauthorized response throws AuthenticationError`() = runTest {
        every { mockResponse.statusCode() } returns 401
        every { mockResponse.body() } returns "Unauthorized"
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        val exception = assertThrows<AuthenticationError> {
            client.getUserInfo("invalid-token")
        }

        assertTrue(exception.message!!.contains("Unauthorized"))
    }

    @Test
    fun `getUserInfo with GraphQL error throws APIError`() = runTest {
        val errorResponse = """{"error": "GraphQL query failed"}"""
        
        every { mockResponse.statusCode() } returns 500
        every { mockResponse.body() } returns errorResponse
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        val exception = assertThrows<APIError> {
            client.getUserInfo("token")
        }

        assertTrue(exception.message!!.contains("GraphQL query failed"))
    }

    @Test
    fun `getUserInfo with fetch error throws APIError`() = runTest {
        val errorResponse = """{"error": "Failed to fetch user info"}"""
        
        every { mockResponse.statusCode() } returns 500
        every { mockResponse.body() } returns errorResponse
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        val exception = assertThrows<APIError> {
            client.getUserInfo("token")
        }

        assertTrue(exception.message!!.contains("Failed to fetch user info"))
    }

    @Test
    fun `getUserInfo with non-success status code throws APIError`() = runTest {
        every { mockResponse.statusCode() } returns 404
        every { mockResponse.body() } returns "Not Found"
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        val exception = assertThrows<APIError> {
            client.getUserInfo("token")
        }

        assertTrue(exception.message!!.contains("HTTP error 404"))
    }

    @Test
    fun `getUserInfo with invalid JSON throws APIError`() = runTest {
        every { mockResponse.statusCode() } returns 200
        every { mockResponse.body() } returns "invalid json"
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        val exception = assertThrows<APIError> {
            client.getUserInfo("token")
        }

        assertTrue(exception.message!!.contains("Failed to parse response"))
    }

    @Test
    fun `getUserInfo sets correct headers`() = runTest {
        val mockResponseBody = """
        {
            "meta": {"code": 200, "message": "Success"},
            "session": {"remainingSeconds": 3600},
            "user": {
                "id": "user123",
                "externalId": "ext123",
                "userName": "testuser",
                "displayName": "Test User",
                "locale": "en-US",
                "active": true,
                "names": {
                    "id": "name123",
                    "familyName": "User",
                    "givenName": "Test"
                },
                "photos": [],
                "phoneNumbers": [],
                "addresses": [],
                "emails": [],
                "verifications": [],
                "provider": "test",
                "createdAt": "2023-01-01T00:00:00Z",
                "updatedAt": "2023-01-01T00:00:00Z",
                "environmentId": "env123"
            }
        }
        """.trimIndent()

        every { mockResponse.statusCode() } returns 200
        every { mockResponse.body() } returns mockResponseBody
        every { 
            mockHttpClient.send(any<HttpRequest>(), any<HttpResponse.BodyHandler<String>>()) 
        } returns mockResponse

        client.getUserInfo("test-token")

        verify {
            mockHttpClient.send(
                match<HttpRequest> { request ->
                    request.uri() == URI.create("https://api.authdog.com/v1/userinfo") &&
                    request.headers().firstValue("Authorization").orElse("") == "Bearer test-token" &&
                    request.headers().firstValue("User-Agent").orElse("") == "authdog-kotlin-sdk/0.1.0" &&
                    request.headers().firstValue("Content-Type").orElse("") == "application/json"
                },
                any<HttpResponse.BodyHandler<String>>()
            )
        }
    }

    @Test
    fun `close method completes without error`() {
        // Should not throw any exception
        client.close()
    }
}
